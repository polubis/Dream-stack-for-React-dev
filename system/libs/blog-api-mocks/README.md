# blog-api-mocks

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build blog-api-mocks` to build the library.

## Running unit tests

Run `nx test blog-api-mocks` to execute the unit tests via [Jest](https://jestjs.io).
