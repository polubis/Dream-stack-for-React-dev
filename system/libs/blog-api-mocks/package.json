{
  "name": "@system/blog-api-mocks",
  "version": "0.0.1",
  "type": "commonjs"
}
