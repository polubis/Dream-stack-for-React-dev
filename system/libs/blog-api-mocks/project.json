{
  "name": "blog-api-mocks",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/blog-api-mocks/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/blog-api-mocks",
        "main": "libs/blog-api-mocks/src/index.ts",
        "tsConfig": "libs/blog-api-mocks/tsconfig.lib.json",
        "assets": ["libs/blog-api-mocks/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/blog-api-mocks/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/blog-api-mocks/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
