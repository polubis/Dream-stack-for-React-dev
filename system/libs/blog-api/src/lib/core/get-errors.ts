import type { ErrorResponse, ResponseError } from '@system/blog-api-models';
import type { AxiosError } from 'axios';

const hasErrors = (
  error: unknown
): error is Required<AxiosError<ErrorResponse>> => {
  const axiosError = <Partial<AxiosError>>error;
  const errorResponse = <Partial<ErrorResponse>>axiosError?.response?.data;

  return (
    !errorResponse?.success &&
    !!errorResponse?.hasErrors &&
    Array.isArray(errorResponse?.errors)
  );
};

const createError = (
  key = 'unknown',
  message = 'Something went wrong...'
): ResponseError => ({
  key,
  message,
});

const getErrors = (error: unknown): ResponseError[] => {
  if (!hasErrors(error)) {
    return [createError()];
  }

  return error.response.data.errors;
};

const getError = (error: unknown): ResponseError => getErrors(error)[0];

export { getError, getErrors, createError };
