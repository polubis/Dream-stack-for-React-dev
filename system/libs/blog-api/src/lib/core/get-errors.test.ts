import { getError, getErrors } from './get-errors';
import type { AxiosError } from 'axios';
import { mockErrorResponse } from '@system/blog-api-mocks';

describe('Errors detection works when: ', () => {
  it('takes errors from response', () => {
    const error = {
      response: {
        data: mockErrorResponse(),
      },
    } as Required<AxiosError>;
    const errors = getErrors(error);

    expect(errors).toEqual(mockErrorResponse().errors);
    expect(getError(error)).toEqual(mockErrorResponse().errors[0]);
  });

  it('returns default errors for other cases', () => {
    const errors = getErrors(mockErrorResponse());

    expect(errors).toMatchSnapshot();
    expect(getError(mockErrorResponse())).toMatchSnapshot();
  });
});
