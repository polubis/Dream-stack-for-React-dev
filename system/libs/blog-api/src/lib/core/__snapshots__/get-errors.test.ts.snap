// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Errors detection works when:  returns default errors for other cases 1`] = `
Array [
  Object {
    "key": "unknown",
    "message": "Something went wrong...",
  },
]
`;

exports[`Errors detection works when:  returns default errors for other cases 2`] = `
Object {
  "key": "unknown",
  "message": "Something went wrong...",
}
`;
